source/mesh_refinement/compaction_length.cc

```diff
@@ -0,0 +1,155 @@
+           cell = this->get_dof_handler().begin_active();
+           cell != this->get_dof_handler().end(); ++cell)
+        {
+          if (cell->is_locally_owned())
+            {
+              bool refine = false;
+              bool clear_coarsen = false;
+
+              fe_values.reinit(cell);
+              in.reinit(fe_values, cell, this->introspection(), this->get_solution(), true);
+              this->get_material_model().evaluate(in, out);
+
+              MaterialModel::MeltOutputs<dim> *melt_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+              AssertThrow(melt_out != NULL,
+                          ExcMessage("Need MeltOutputs from the material model for computing the melt properties."));
+
+              // for each composition dof, check if the compaction length exceeds the cell size
+              for (unsigned int i=0; i<this->get_fe().base_element(this->introspection().base_elements.compositional_fields).dofs_per_cell; ++i)
+                {
+                  const double porosity = in.composition[i][porosity_idx];
+                  const double compaction_length = std::sqrt((out.viscosities[i] + 4./3. * melt_out->compaction_viscosities[i])
+                                                             * melt_out->permeabilities[i] / melt_out->fluid_viscosities[i]);
+
+                  // If the compaction length exceeds the cell diameter anywhere in the cell, cell is marked for refinement.
+                  // Do not apply any refinement if the porosity is so small that melt can not migrate.
+                  if (compaction_length < 2.0 * cells_per_compaction_length * cell->minimum_vertex_distance()
+                      && porosity > this->get_melt_handler().melt_transport_threshold)
+                    clear_coarsen = true;
+
+                  if (compaction_length < cells_per_compaction_length * cell->minimum_vertex_distance()
+                      && porosity > this->get_melt_handler().melt_transport_threshold)
+                    {
+                      refine = true;
+                      break;
+                    }
+                }
+
+              if (clear_coarsen)
+                cell->clear_coarsen_flag ();
+              if (refine)
+                cell->set_refine_flag ();
+            }
+        }
+    }
+
+    template <int dim>
+    void
+    CompactionLength<dim>::
+    declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Compaction length");
+        {
+          prm.declare_entry("Mesh cells per compaction length", "1.0",
+                            Patterns::Double (0),
+                            "The ratio between compaction length and size of the mesh cells, "
+                            "or, in other words, how many cells the mesh should haver per "
+                            "compaction length.");
```

Comment:
I think you should explain in this description how this parameter is used, e.g., "This mesh refinement criterion will mark cells for refinement if ..."